### ISSUE TITLE
[Suggestion] Auto tracking C++11 dependencies

### ISSUE THREAD

**Issue by StevenPuttemans** [opencv/opencv_contrib#1122]
I was trying to get the SFM module to built. For this the ceres solver is needed. So the first step was to built the ceres-solver. In the CMAKE, ceres allows users to use C++11 defined directives to improve the performance. This is not activated by default, but it allows users to make use of the full potential of their machine.

However, when OpenCV CMAKE auto detects its ceres dependency, it does not check if this is built with C++11 support or not. _(not even aware if this is possible to do)_ 

So when building OpenCV with contrib modules then, SFM module is enabled, but when building a tons of errors occur. A small google gave me the hint that C++11 directives were not found and that I had to add the specific `-D CMAKE_CXX_STANDARD=11` entry to CMAKE, since CMAKE still does not select C++11 by default.

Is there a possibility to add this flag by default to the CMAKE layout of OpenCV? Or somehow warm people that if dependencies of OpenCV are built with C++11 support, that OpenCV needs to be built with it also?

**Comment by sergiud** [comment]
This is not an OpenCV issue. Rather, Ceres must indicate in its package config that it uses C++11 features using `target_compile_features` or by setting the `INTERFACE_COMPILE_FEATURES` property.

See https://cmake.org/cmake/help/v3.8/command/target_compile_features.html and https://cmake.org/cmake/help/v3.8/prop_tgt/INTERFACE_COMPILE_FEATURES.html.

**Comment by StevenPuttemans** [comment]
@sergiud I will reference this issue at the ceres issue tracker, asking them to add this directive.

**Comment by StevenPuttemans** [comment]
Okay an update on the matter using the feedback of the guys from the ceres-solver repository.

Ceres does provide a way of telling that it was compiled with C++ through `INTERFACE_COMPILE_OPTIONS` which is also available in older CMAKE versions, because `target_compile_features()` only arrived in CMAKE3.1.

It can also be checked if CERES was compiled with C++11 by checking `CERES_COMPILED_COMPONENTS` variable defined after a call to `find_package(Ceres)`.

It also seems OpenCV is able to find Ceres and its C++11 pipe during cmake, knowing that the output generates 

```
Ceres version 1.12.0 detected here: /usr/local was built with C++11. Ceres target will add C++11 flags to compile options for targets using it.
Found Ceres version: 1.12.0 installed in: /usr/local with components: [LAPACK, SuiteSparse, SparseLinearAlgebraLibrary, CXSparse, SchurSpecializations, C++11, OpenMP]
```

After providing that as a feedback to the guys of the ceres-solver, they replyed with

> @StevenPuttemans your problem actually looks like a bug in the OpenCV SfM CMake. Looking at the latest master branch, you're compiling fundamental.cc which is part of the multiview library, however despite (optionally) using Ceres that library never declares that it is dependent upon Ceres via target_link_libraries(). It apparently only depends on the 'numeric' library, which does not depend on Ceres.
> 
> The 'simple_pipeline' library does declare that it uses Ceres, and so the '-std=c++11' flag from the Ceres target should be being appended to that target's compile definitions, but this will not recurse back up the tree of dependencies to the 'multiview' library.
> 
> Given this issue, I don't think that Ceres using target_compile_features() would actually solve your problem. If you fix the OpenCV SfM CMake, the current solution should work for you. However, in general I would advise against compiling OpenCV without C++11 if Ceres is compiled with it - although the language is ABI compatible, parts of the STL are not and mixing them can be problematic. YMMV, but in general it is not advisable.

So I guess we will have to get the CMAKE of the SfM module in order.

**Comment by johncant** [comment]
Hi - I think I've implemented the above

> So I guess we will have to get the CMAKE of the SfM module in order.

Is there anything else you have in mind WRT the SfM module CMake? I am happy to take a look. I also noticed while trying to build with minimal modules that `libmv_light` inside `sfm` has unspecified dependencies on `highgui` and `flann`, by using `cv::imread` and `cv::FlannBasedMatcher`. Also happy to backport fixes to opencv 2.4.
